*Creating A Feedforward Neural Network In PyTorch To Predict Earthquake Depth Ranges*


Import libraries

In [130]:
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from imblearn.over_sampling import SMOTE
import pandas as pd
import numpy as np

Load the cleaned csv file.

In [131]:
# Load dataset
clean_eq = pd.read_csv('/Users/gerryjr/Desktop/EarthquakeProject/Data/clean_earthquake.csv')

Before getting into the process of filling/cleaning any missing data, let's create ranges(bins) to categorize earthquake depths. Concurrently, let's map the "labels" which are strings. This gives the strings of "shallow", "intermediate" and "deep" a numeric label of 0, 1 and 2. Since this will be a Feedforward Neural Network, the model needs the inputs to be numeric.

In [132]:
# Create depth ranges
clean_eq['depth_range'] = pd.cut(clean_eq['depth'], bins=[0, 70, 300, float('inf')], labels=['Shallow', 'Intermediate', 'Deep'])

# Map depth ranges to numeric labels
depth_mapping = {'Shallow': 0, 'Intermediate': 1, 'Deep': 2}
clean_eq['depth_range'] = clean_eq['depth_range'].map(depth_mapping)

Next, this drops any rows in the "depth_range" column that has missing or non-numeric values. This will be important for the model.

In [133]:

clean_eq = clean_eq.dropna(subset=['depth_range'])

Just like the TensorFlow FNN, features are engineered. The follow feature engineering is related to depth which will be important for the model to train from.

In [134]:
#Feature Engineering
clean_eq['depth_squared'] = clean_eq['depth'] ** 2
clean_eq['depth_error_ratio'] = clean_eq['depthError'] / clean_eq['depth']
clean_eq['depth_error_ratio'].replace([np.inf, -np.inf], np.nan, inplace=True)
clean_eq['depth_error_ratio'].fillna(clean_eq['depth_error_ratio'].median(), inplace=True)
clean_eq['lat_long_interaction'] = clean_eq['latitude'] * clean_eq['longitude']


/var/folders/_2/d_jf9l7x2rl2fp7yfmdp4c100000gn/T/ipykernel_3742/2690894172.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  clean_eq['depth_error_ratio'].replace([np.inf, -np.inf], np.nan, inplace=True)
/var/folders/_2/d_jf9l7x2rl2fp7yfmdp4c100000gn/T/ipykernel_3742/2690894172.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on whic

The features are listed with the fixed and engineered variants. In this case, the target is "depth_range."

In [135]:
# Features and target
X = clean_eq[['depth',
              'gap',
              'rms',
              'depth_squared',
              'latitude',
              'longitude',
              'depth_error_ratio',
              'lat_long_interaction']]
y = clean_eq['depth_range']





Like other models, this splits the testing and training data with 20% being set aside for testing.

In [136]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
X_train = X_train.fillna(X_train.mean())  # or .median()
X_test = X_test.fillna(X_test.mean())


Since this is a neural network, it is important to scale the features.

In [137]:
# Scale features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

The below cell implenets SMOTE which stand sfor Synthetic Minority Over-Sampling Technique. This allows the model to deal with class imbalances in the data. In the previous Jupyter Notebook you will notice the distribution plot showed am imbalance of earthquakes of magnitudes ~5. This means that the model can encounter an issue with identifying outliers (aka minorities). I added this cell afterwards when the model was struggling to identify intermidiate and deep depth ranges. This will become more apparent when the results are shown for the model performance.

In [138]:
# Apply SMOTE
smote = SMOTE(sampling_strategy='auto', random_state=42)
X_train, y_train = smote.fit_resample(X_train, y_train)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/base.py:474: FutureWarning: `BaseEstimator._validate_data` is deprecated in 1.6 and will be removed in 1.7. Use `sklearn.utils.validation.validate_data` instead. This function becomes public and is part of the scikit-learn developer API.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/utils/_tags.py:354: FutureWarning: The SMOTE or classes from which it inherits use `_get_tags` and `_more_tags`. Please define the `__sklearn_tags__` method, or inherit from `sklearn.base.BaseEstimator` and/or other appropriate mixins such as `sklearn.base.TransformerMixin`, `sklearn.base.ClassifierMixin`, `sklearn.base.RegressorMixin`, and `sklearn.base.OutlierMixin`. From scikit-learn 1.7, not defining `__sklearn_tags__` will raise an error.
  warnings.warn(


While this looks a bit abstract, these 4 lines simply tell the model that 32-bit floating are the inputs. 32-floating point numbers are standard procedure for deep learning models.

In [139]:
# Convert to PyTorch tensors
X_train = torch.tensor(X_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_train = torch.tensor(y_train.values, dtype=torch.long)
y_test = torch.tensor(y_test.values, dtype=torch.long)

Now it is time to create the FNN. Since this is utilizing PyTorch the syntax. Since PyTorch runs the computations dynamically, the code is slightly more rigorous than TensorFlow. However, the overall methodology does not change. A dropout of 40% is added to force the model to learn patterns instead of memorizing the data. The "relu" activation function is added for non-linearity. 

In [140]:
# FNN with dropout
class FNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(FNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(0.4)  # Dropout to reduce overfitting
        self.fc2 = nn.Linear(hidden_size, output_size)
    
    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.dropout(x)
        x = self.fc2(x)
        return x

Initialize the function with the layers and size of said layers.

In [141]:
# Initialize model
input_size = X_train.shape[1]
hidden_size = 128  # Adjusted hidden layer size
output_size = len(depth_mapping)
model = FNN(input_size, hidden_size, output_size)

The loss and optimization are defined. This measures how accurate the predictions are and modifies how the model adjusts the neuron weights during the backpropagation portion of the algorithm. 

In [142]:
# Loss function and optimizer 
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001, weight_decay=0.001)  # L2 regularization


Now it is time to train the model. The iterations (epochs) are set as 50. This is written as a loop for each iteration. Again PyTorch computes dynamically so there is more flexibility in the structure of the algorithm. 

In [143]:
# Training the FNN
num_epochs = 50
for epoch in range(num_epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()
    
    if (epoch + 1) % 10 == 0:
        print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")

Epoch [10/50], Loss: 0.7998
Epoch [20/50], Loss: 0.6824
Epoch [30/50], Loss: 0.5929
Epoch [40/50], Loss: 0.5238
Epoch [50/50], Loss: 0.4664


Since the model is trained, now it is important to evaluate how well it performed. This prints the accuracy and a classification report. This will discriminate the 3 sub-targets(shallow, intermediate and deep) to evaluate how accurately the model predicted them. To have a visual correlation, a "Confusion Matrix" is generate to show how well the model did.

In [144]:
# Evaluate the model
model.eval()
with torch.no_grad():
    test_outputs = model(X_test)
    _, predicted = torch.max(test_outputs, 1)
    accuracy = accuracy_score(y_test, predicted)
    print(f"Test Accuracy: {accuracy:.4f}")
    print("Classification Report:")
    print(classification_report(y_test, predicted, target_names=['Shallow', 'Intermediate', 'Deep']))

    #Confusion matrix
    cm = confusion_matrix(y_test, predicted)
    print("Confusion Matrix:")
    print(cm)


Test Accuracy: 0.9565
Classification Report:
              precision    recall  f1-score   support

     Shallow       0.98      0.97      0.98      1271
Intermediate       0.83      0.81      0.82       189
        Deep       0.95      1.00      0.98        80

    accuracy                           0.96      1540
   macro avg       0.92      0.93      0.92      1540
weighted avg       0.96      0.96      0.96      1540

Confusion Matrix:
[[1239   32    0]
 [  31  154    4]
 [   0    0   80]]


The above output indicates the model is performing quite well, consistently performing in the low to middle 90%. The "shallow" class performed at 98% which is very strong. The intermidiate class had some missclassifications at 70%. This is somewhat expected since the model is conflating scenarios that are close to the "deep" class. Finally, the "deep" class performed at 88% which is respectable. Overall the model performs well.